### PR TITLE
Prebuilt RectSlice

### PR DESCRIPTION
**Pull request by namse**
Design Doc : https://docs.google.com/document/d/1HlSZHpqZlAomAh3q0eM60ppCpepoJTQyM4atJk282XA/edit#

# What is RectSlice?

RectSlice is grid system like Fluid Container of MUI.

It takes everything in rectangular mood. It slices body vertically or horizontally.

You can slice RectSlice, then RectSlice will be sliced as 2 parts, A and B.
one of A or B has fixed(or parent-size-relative) size. And the other part has flexible size that fills whole rest space.

example
```
[ A: 20px | B: rest of parent! ]
```

You can put RenderingTree or other RectSlice recursively inside of A or B. So, you can make **grid** or **table** using RectSlice.

### CODE REVIEW

File: namui-prebuilt/src/dubu_gat.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
+// NOTE: This is unused file for now. I will update it as main when GAT is ready on rust.
```
Comment (namse) [inline namui-prebuilt/src/dubu_gat.rs:1]: Please do not review this code yet. I will use this code later when GAT is ready for rust.

File: namui-prebuilt/src/dubu.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    pub fn render(&self, props: Props) -> RenderingTree {
+        self.dubu.render(self.wh, props)
+    }
+}
+
+mod test {
+    use super::*;
+
+    #[allow(dead_code)]
+    fn example() -> RenderingTree {
```
Comment (namse) [inline namui-prebuilt/src/dubu.rs:136]: image

File: namui-prebuilt/src/rect_slice/mod.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+#[cfg(test)]
+mod tests {
+    use self::traits::RectSlice;
+    use super::*;
+    use std::sync::atomic::AtomicBool;
+    use wasm_bindgen_test::wasm_bindgen_test;
+
+    #[test]
+    #[wasm_bindgen_test]
+    fn layout_should_give_right_wh() {
```
Comment (namse) [inline namui-prebuilt/src/rect_slice/mod.rs:130]: image

File: namui-prebuilt/src/rect_slice/mod.rs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        assert_eq!(
+            true,
+            body_render_called.load(std::sync::atomic::Ordering::Relaxed)
+        );
+    }
+
+    #[test]
+    #[wasm_bindgen_test]
+    fn layout_with_traits_should_give_right_wh() {
+        let label_render_called = AtomicBool::new(false);
```
Comment (namse) [inline namui-prebuilt/src/rect_slice/mod.rs:192]: image

File: namui-prebuilt/src/rect_slice/mod.rs (hunk 3 above)
Comment (bigfoodK) [reply namui-prebuilt/src/rect_slice/mod.rs:130]: *컨테이너 (300px, 500px)

Comment (namse) [reply namui-prebuilt/src/rect_slice/mod.rs:130]: Ah right

**Review by bigfoodK — APPROVED** [review]

File: namui-prebuilt/src/rect_slice/mod.rs [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        );
+        assert_eq!(
+            true,
+            body_render_called.load(std::sync::atomic::Ordering::Relaxed)
+        );
+    }
+
+    #[test]
+    #[wasm_bindgen_test]
+    fn layout_with_traits_should_give_right_wh() {
```
Comment (bigfoodK) [inline namui-prebuilt/src/rect_slice/mod.rs:191]: I like trait-way.

Comment (namse) [reply namui-prebuilt/src/rect_slice/mod.rs:191]: You can use trait-way but if you want to use simple things without struct, then struct-way will be very helpful.

**Pull request merged**